src/nat_test.c

```diff
@@ -0,0 +1,545 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <cJSON.h>
+#include <cJSON_os.h>
+#include <modem/lte_lc.h>
+#include <modem/modem_info.h>
+#include <net/socket.h>
+#include <stdarg.h>
+#include <stdio.h>
+
+#include "nat_test.h"
+
+#define UDP_PORT 3050
+#define TCP_PORT 3051
+#define THREAD_STACK_SIZE 8192
+#define WAIT_TIME_S 3
+#define TIMEOUT_TOL_S 10
+#define DEFAULT_UDP_INITIAL_TIMEOUT 1
+#define DEFAULT_TCP_INITIAL_TIMEOUT 300
+#define DEFAULT_UDP_TIMEOUT_MULTIPLIER 2
+#define DEFAULT_TCP_TIMEOUT_MULTIPLIER 1.5
+#define IP_STRINGS_COUNT 10
+
+struct test_thread_timeout {
+	int timeout;
+	double multiplier;
+	int lower;
+	int upper;
+};
+
+struct test_thread_data {
+	atomic_t type;
+	atomic_t state;
+	struct k_sem sem;
+	struct test_thread_timeout timeout_data;
+};
+
+struct test_thread {
+	struct k_thread thread;
+	k_tid_t tid;
+	k_thread_stack_t *stack_area;
+	struct test_thread_data thread_data;
+};
+
+K_THREAD_STACK_DEFINE(nat_test_thread_stack_area, THREAD_STACK_SIZE);
+
+static struct test_thread test_thread;
+
+volatile int udp_initial_timeout;
+volatile int tcp_initial_timeout;
+volatile float udp_timeout_multiplier;
+volatile float tcp_timeout_multiplier;
+
+int get_test_state(void)
+{
+	return atomic_get(&test_thread.thread_data.state);
+}
+
+static int json_add_obj(cJSON *parent, const char *str, cJSON *item)
+{
+	cJSON_AddItemToObject(parent, str, item);
+
+	return 0;
+}
+
+static int json_add_str(cJSON *parent, const char *str, const char *item)
+{
+	cJSON *json_str;
+
+	json_str = cJSON_CreateString(item);
+	if (json_str == NULL) {
+		return -ENOMEM;
+	}
+
+	return json_add_obj(parent, str, json_str);
+}
+
+static int json_add_number(cJSON *parent, const char *str, double item)
+{
+	cJSON *json_num;
+
+	json_num = cJSON_CreateNumber(item);
+	if (json_num == NULL) {
+		return -ENOMEM;
+	}
+
+	return json_add_obj(parent, str, json_num);
+}
+
+static int create_send_buffer(struct modem_param_info *modem_params,
+			      char *buffer, int timeout_s)
+{
+	int ret = 0;
+	cJSON *root_obj = cJSON_CreateObject();
+	cJSON *ip_obj;
+	const char *delim = " ";
+	const char *ip_strings[IP_STRINGS_COUNT];
+	int ip_count = 0;
+
+	if (root_obj == NULL) {
+		printk("Failed to create json root object\n");
+		return -ENOMEM;
+	}
+
+	char *token =
+		strtok(modem_params->network.ip_address.value_string, delim);
+	while (token != NULL) {
+		if (ip_count >= ARRAY_SIZE(ip_strings)) {
+			printk("More than %d addresses found. Remainder will not be added to json\n",
+			       IP_STRINGS_COUNT);
+			break;
+		}
+		ip_strings[ip_count] = token;
+		token = strtok(NULL, delim);
+		ip_count++;
+	}
+
+	ip_obj = cJSON_CreateStringArray(ip_strings, ip_count);
+	if (ip_obj == NULL) {
+		printk("Failed to create json ip object\n");
+		return -ENOMEM;
```

Comment:
When returning here, `root_obj` well not be freed, so there's a memory leak. It should instead `goto exit`

Reply:
Fixed